src/app/left-menu/main/left-menu.component.ts

```diff
@@ -52,23 +53,112 @@ export class LeftMenuComponent implements OnDestroy, OnInit {
         e.stopPropagation();
     }
 
+    changeColor(text: string) {
+       const keyword = this.getRealName(this.searchKeyword.toLowerCase());
+       if (keyword.length >= 3) {
+        text = '<span style="color:#f05b41">' + text.slice(0, keyword.length) + '</span>' + text.slice(keyword.length);
+        return this.sanitizer.bypassSecurityTrustHtml(text);
+       }
+       return text;
+    }
+
     menuSearch() {
-        const keyword = this.searchKeyword.toLowerCase();
-        this.filteredMenuData = this.menuData.filter((value) => {
-            const name = value.name.toLowerCase();
-            const searchString = name + ', ' + value.equivalents;
-            return searchString.toLowerCase().indexOf(keyword) !== -1;
-        });
+        const keyword = this.getRealName(this.searchKeyword.toLowerCase());
+
+        if (keyword.length < 3) {
+            this.filteredData = [];
+            this.filteredData[0] = this.workArea;
+        }
+
+        if (keyword.length >= 3) {
+            this.filteredData = this.deepClone(this.menuData);
```

Comment:
You approach:
get copy of full data - filter the copy - show filtered copy
Suggested:
get filtered from full - show 

In other words
this.filteredData = this.menuData.filter(superFilteredLogic);

This approach does not keep an extra full copy of data in memory and does not need `deepCopy` method.

Reply:
The problem is menuData. When I change filteredData menuData also changes because it is object. So i use deepCopy method.

Reply:
You do not need to change menuData and copy it to filteredData at the start. Just collect what you need from menuData and copy this to filteredData. And, yes, in this case, you do not need to clear empty objects. 
```
// pseudocode (I do not remember the names of all fields)
this.filteredData = [];
menuData.forEach((menuDataItem) => {
    const filteredDataItem = {}; // maybe it is array, I do not know
    menuDataItem.groups.forEach((menuDataItemGroup) => {
         if(whatINeed1(menuDataItemGroup)) {
             filteredDataItem .push(menuDataItem);
         }
    });
    if(whatINeed2(menuDataItem)) {
         this.filteredData.push(menuDataItem);
    }
    
});
``` 
Be closer to real life. You do not need to bring at home all melons from the market to select a single one.